src/utils/flagging.R

```diff
@@ -0,0 +1,306 @@
+    }
+
+    # check the lengths of the inputs
+    ln_p <- length(periods)
+    ln_tp <- length(thresholds_pcts)
+    ln_ts <- length(thresholds_static)
+
+
+    if ((ln_p != ln_tp && ln_tp != 1) || (ln_p != ln_ts && ln_ts != 1)) {
+      stop(
+        "`thresholds_pcts` and `thresholds_static` must have the same length ",
+        "as `periods` if passed to `calculate_alerts()`, unless length 1, in ",
+        " which case they are recycled.",
+        call. = FALSE
+      )
+    }
+
+    # reorderig for later pmap
+    if (ln_tp > 1) {
+      thresholds_pcts <- thresholds_pcts[p_order]
+    }
+
+    if (ln_ts > 1) {
+      thresholds_static <- thresholds_static[p_order]
+    }
+
+    if (is.null(thresholds_minimums)) {
+      thresholds_minimums <- 0
+    } else {
+      ln_m <- length(thresholds_minimums)
+      if (ln_p != ln_m && (ln_m != 1)) {
+        stop(
+          "`thresholds_minimum` must have the same length ",
+          "as `periods` if passed to `calculate_alerts()`, unless length 1, in ",
+          " which case it is recycled.",
+          call. = FALSE
+        )
+      } else if (ln_m > 1) {
+        thresholds_minimums <- thresholds_minimums[p_order]
+      }
+    }
+  }
+
+  # create the rolling sum columns necessary for flagging
+  rs_periods <- c(first_since, periods)
+  for (period in rs_periods) {
+    .data <- mutate(
+      .data,
+      "rs_{{ period }}" := zoo::rollsumr(
+        x = {{ x }},
+        k = !!period,
+        fill = NA
+      )
+    )
+  }
+
+  # if first since is passed
+  if (!null_fs) {
+    first_since <- sort(first_since, decreasing = FALSE)
+    for (period in first_since) {
+      .data <- mutate(
+        .data,
+        "flag_first_{{ period }}" := {{ x }} > 0 & lag(.data[[paste0("rs_", period)]]) == 0
+      )
+    }
+  }
+
+  # if period is passed
+  if (!null_p) {
+    new_df <- pmap(
+      .l = list(
+        periods,
+        thresholds_pcts,
+        thresholds_static,
+        thresholds_minimums
+      ),
+      .f = \(period, pct, threshold, minimum) {
+        mutate(
+          .data,
+          "flag_anomaly_{{ period }}" := flag_fun(
+            x = .data[[paste0("rs_", period)]],
+            pct = pct,
+            threshold = threshold,
+            minimum = minimum
+          )
+        ) %>%
+          ungroup() %>%
+          select(
+            starts_with("flag_anomaly_")
+          )
+      }
+    ) %>%
+      list_cbind()
+
+    .data <- bind_cols(.data, new_df) %>%
+      mutate(
+        across(
+          .cols = starts_with("flag_"),
+          .fns = \(x) replace_na(x, FALSE)
+        )
+      )
+  }
+
+  # find the flag levels for each time point
+  # the higher the level, the more critical the flag
+
+  .data$alert_levels <- .data %>%
+    ungroup() %>%
+    select(
+      starts_with("flag_")
+    ) %>%
+    data.matrix() %>%
+    apply(
+      MARGIN = 1,
+      FUN = \(x) max(which(x == 1), -Inf)
+    ) %>%
+    pmax(
+      0
+    )
```

Comment:
I've added comments to make it clear what we're doing. The speed increases of using matrix algebra here is much quicker than other approaches. All we are doing is simply identifying in which column is the furtherest right that has a positive flag. This only works if we have properly ordered our flags, which we have done. Thus, all we do is:

1. select all the flag columns
2. convert to a matrix
3. find for each row, which is the last column with a positive flag

Then we have a numeric vector of flags that increases as the flag severity increases.

Reply:
True that matrix algebra is faster, but don't think speed is a primary concern here- especially since the alternative takes ~1s to run. But matrix algebra is cool and powerful, so maybe we should use it more anyways :-)

With comments i think it is fine either way - just want to ensure someone else can interpret the code w/ out to much up-front work.